knitr/hmm-example/hmm-example.Rmd

```diff
@@ -0,0 +1,269 @@
+To make it clear how to use the HMM fit functions, the model here will fit the
+transition matrix, the initial state, and the parameters of the measurement
+model. It is not necessary to estimate all of these things in practice if some
+of them are known.
+
+The data is the previously generated sequence of $N$ measurements:
+```{stan, output.var = "", eval = FALSE}
+data {
+  int N; // Number of observations
+  real y[N];
+}
+```
+
+For the transition matrix, assume that it is known that states 1 and 3 are not
+directly connected. For $K$ states, estimating a full transition matrix means
+estimatng a matrix of $O(K^2)$ probabilities. Depending on the data available,
+this may not be possible and so it is important to take advantage of available
+modeling assumptions. The state means are estimated as an ordered vector
+to avoid mode-swap non-identifiabilities.
+
+```{stan, output.var = "", eval = FALSE}
+parameters {
+  // Rows of the transition matrix
+  simplex[2] t1;
+  simplex[3] t2;
+  simplex[2] t3;
+  
+  // Initial state
+  simplex[3] rho;
+  
+  // Parameters of measurement model
+  vector[3] mu;
+  real<lower = 0.0> sigma;
+}
+```
+
+The `hmm_marginal` function takes the transition matrix and initial state
+directly. In this case the transition matrix needs constructed from `t1`,
+`t2`, and `t3` but that is relatively easy to build.
+
+The measurement model, in contrast, is not passed directly to the HMM function.
+
+Instead, a $KxN$ matrix `log_omega` of log likelihoods is passed in. The
+$(k, n)$ entry of this matrix is the log likelihood of the $nth$ measurement
+given the system at time $n$ is actually in state $k$. For the generative
+model above, these are log normals evaluated at the three different means.
+
+```{stan, output.var = "", eval = FALSE}
+transformed parameters {
+  matrix[3, 3] gamma = rep_matrix(0, 3, 3);
+  matrix[3, N] log_omega;
+  
+  // Build the transition matrix
+  gamma[1, 1:2] = t1;
+  gamma[2, ] = t2;
+  gamma[3, 2:3] = t3;
+  
+  // Compute the log likelihoods in each possible state
+  for(n in 1:N) {
+    // The observation model could change with n, or vary in a number of
+    //  different ways (which is why log_omega is passed in as an argument)
+    log_omega[1, n] = normal_lpdf(y[n] | mu[1], sigma);
+    log_omega[2, n] = normal_lpdf(y[n] | mu[2], sigma);
+    log_omega[3, n] = normal_lpdf(y[n] | mu[3], sigma);
+  }
+}
+```
+
+With all that in place, the only thing left to do is add priors and increment
+the log density:
+```{stan, output.var = "", eval = FALSE}
+model {
+  mu ~ normal(0, 1);
+  sigma ~ normal(0, 1);
+
+  target += hmm_marginal(log_omega, Gamma, rho);
+}
+```
+
+```{r echo = TRUE, results = FALSE, message = FALSE}
+model = cmdstan_model("hmm-example.stan")
+fit = model$sample(data = list(N = N, y = y), parallel_chains = 4)
+```
+
+The estimated group means match the known ones:
+```{r}
+fit$summary("mu")
+```
+The estimated of the initial condition is not much more informative than
```

Comment:
Either the "estimates of the initial condition" or the "estimated initial condition".